### PR TITLE
ssbc: Persist cache result on workspace

### PR DESCRIPTION
**Pull request by eseliger**
This PR fixes https://github.com/sourcegraph/sourcegraph/issues/28080. 

Before, we would lose cached status on steps when the cache is evicted. Also, when using the `noCache` option, we would still show step cache results briefly in the UI. This PR also enforces that we have step cache results available for all steps before the most recent cache result (before, we could have a result for step 3, but none for 1 and 2 and those would render incorrectly and have no diff etc).

The downside:
This duplicates a lot of data. This might become problematic in the future. This is a really easy and straightforward fix and I don't anticipate anyone before GA running into big trouble here, but if we think this is a big blocker, we can address this now as well.
An alternative solution would be to only store the link between step and cache result, as a foreign key. We could add a table workspace_step_cache_entries (workspace_id, cache_entry_id).

Then we would need to adjust our pruning algorithm:
- We don't count referenced cache entries in the total cache size when we calculate what to prune
- We do allow duplicates in the key column
  - and when reading cache entries for use, we do `SELECT max(id) .. GROUP BY key`
  - when inserting a new cache entry we
    - insert as before when there's no collision
    - update the record in place, when there are 0 references to this entry
    - create a new record

If we decide to proceed with this approach _for now_, because it fixes the issue at some additional storage cost, I'll put this PR description in a ticket for post beta work. Otherwise, I'll pick this up again and incorporate these suggestions into this PR. I don't have a strong preference but would err on the side of moving fast.

### CODE REVIEW

File: enterprise/internal/batches/background/batch_spec_workspace_creator.go [hunk 1]
```diff
@@ -155,14 +174,42 @@ func (r *batchSpecWorkspaceCreator) process(
...
 
 	// Check for an existing cache entry for each of the workspaces.
 	for rawKey, workspace := range cacheKeyWorkspaces {
+		for idx, key := range workspace.stepCacheKeys {
+			if c, ok := stepEntriesByCacheKey[key]; ok {
+				var res execution.AfterStepResult
+				if err := json.Unmarshal([]byte(c.Value), &res); err != nil {
+					return err
+				}
+				workspace.dbWorkspace.SetStepCacheResult(idx+1, btypes.StepCacheResult{Key: key, Value: &res})
```
Comment (mrnugget) [inline enterprise/internal/batches/background/batch_spec_workspace_creator.go:205]: Is there a reason for the idx+1 convention? It seems easier to say that `idx == 0` means "cached results _after_ step 0" everywhere. With the current implementation the iterating over steps is always 0-based but the cache-entries themselves are 1-based.

File: enterprise/internal/batches/background/batch_spec_workspace_creator.go [hunk 2]
```diff
@@ -124,27 +125,45 @@ func (r *batchSpecWorkspaceCreator) process(
...
 			w.Steps,
 		)
+
 		rawKey, err := key.Key()
 		if err != nil {
 			return err
 		}
+
+		stepCacheKeys := make([]string, 0, len(workspace.Steps))
+		// Generate cache keys for all the step results as well.
```
Comment (mrnugget) [inline enterprise/internal/batches/background/batch_spec_workspace_creator.go:135]:
```suggestion
		// Generate cache keys for all the step results, except the last one.
```

Right?

**Review by mrnugget — APPROVED** [review]

File: enterprise/internal/batches/background/batch_spec_workspace_creator.go (hunk 1 above)
Comment (eseliger) [reply enterprise/internal/batches/background/batch_spec_workspace_creator.go:205]: At some point we made the mistake that steps are 1-indexed, and I didn't want to make it even worse by also mixing the two up here. I think we should ultimately fix all of this in one go. WDYT?

File: enterprise/internal/batches/background/batch_spec_workspace_creator.go (hunk 2 above)
Comment (eseliger) [reply enterprise/internal/batches/background/batch_spec_workspace_creator.go:135]: No, this is a 0..N-1 loop so it still created it for all entries. This is the only case where we use 0-index 🥼 SORRY :handcuffs:

File: enterprise/internal/batches/background/batch_spec_workspace_creator.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/batches/background/batch_spec_workspace_creator.go:205]: But where are they 1-indexed? The step results, for example, are 0-indexed:

https://github.com/sourcegraph/sourcegraph/blob/46d2566d16841939f3271a0877f0cd1430d056ac/lib/batches/execution/results.go#L51-L55

Do you mean the parsing of JSON log liens? https://github.com/sourcegraph/sourcegraph/blob/dab2adbfcaac1f9aad8e836bdcabc07e4bfed845/enterprise/internal/batches/types/step_info.go#L51:6

**Pull request merged**